core/sdk-core/src/main/java/software/amazon/awssdk/core/client/handler/BaseClientHandler.java

```diff
@@ -52,7 +55,9 @@ protected BaseClientHandler(SdkClientConfiguration clientConfiguration) {
 
         runBeforeMarshallingInterceptors(executionContext);
         Request<InputT> request = executionParams.getMarshaller().marshall(inputT);
-        request.setEndpoint(clientConfiguration.option(SdkClientOption.ENDPOINT));
+
+
+        request.setEndpoint(resolveEndpoint(clientConfiguration, executionParams.hostPrefixExpression()));
```

Comment:
This actually changes a bit with the marshaller refactor. I'm curious to get your opinion on this, right now I'm passing the client configured endpoint to the marshaller and setting the endpoint there. I need to do that because SdkHttpFullRequest can't be built without an endpoint. We could a couple of things I think

1) Relax constraints in SdkHttpFullRequest to not require endpoint on build so we can open it back up later and add endpoint
2) Change marshallers to produce a SdkHttpFullRequest.Builder and fill in the endpoint and build later on in the request lifecycle
3) Somehow get the operation level endpoint to the marshaller and set it there
4) Continue to just set the client level endpoint knowing that it may be overwritten later

Thoughts?

Reply:
I am leaning towards 4.
> 2. Change marshallers to produce a SdkHttpFullRequest.Builder and fill in the endpoint and build later on in the request lifecycle

I think its better to pass built objects to avoid any mutability concerns

> 3. Somehow get the operation level endpoint to the marshaller and set it there

Have to do it for each protocol. Don't like it

> 4. Continue to just set the client level endpoint knowing that it may be overwritten later

I like this as we get marshalled request and we can construct a new request (with new endpoint) if host prefix is set. As we don't reuse the marshalled request at any place other than `BaseClientHandler`, it seems safe and modification happen at single place. 
If we are choosing between 4 and 1, relaxing constraints doesn't give any advantage and 4 seems ideal.

What do you think?

Reply:
Discussed offline. Going to keep the logic to update endpoint in `BaseClientHandler` instead of moving to marshallers as we need to pass more information every time we need to support a feature updating endpoint (like endpoint discovery).